### ISSUE TITLE
Incorrect TS7022 error in loop

### ISSUE THREAD

**Issue by mysticatea** [microsoft/TypeScript#42529]
# Bug Report

Tsc threw TS7022 error on the place that is not `any` type.

```ts
        const group = await fetchGroup(nextId); // (id: string) => Promise<Group>
        //    ^^^^^ 'group' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. (7022)
```

### 🔎 Search Terms

- `7022`

#33191 and #39183 are maybe related. But in this case, the return type of `fetchGroup()` does not depend on parameters; that always `Promise<Group>` explicitly. Also, the type of `nextId` is always `string` at the line.

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about 7022

(3.3.3 ... 4.2.0-dev.20210127)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=20&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgOJQPYFcAOyDeAUMicjnFBOAJIAmAXMgM5hSgDmxpA9NwHQDCAX0KFaEBABsKKGFhAIwwDCGQwIYBAAt02HAApgDZqw4BKRgAVMAW2BMIAHl24AfKLhMAngrXzFyqrqmloAggoQLBhQLjhM+pIQcDB0jCxsIOwWyNYYdg7OmLgA2gC6rgRcJAgqLMjsRXGMsWXIALzIZQDcVciJYMggEAAeYKkmGezIAD7I8uIwoBC07X1JKbQ9vQDuWsCJyPpDo3RmlaQXyDUgdQ16q3DbcMADwdqxRyNjtGY9lzzcC4APRBIOQAHI7rhwchgDYcJJgAgXpIvMgtJ5kGAvDgUOC4CAvDCAEYSOBYBywga0DCRQYYAYYgBuKDgWJxrJAIAZcCUKmQBJW9mQlHUlAiK1owEoilRyGisJAUplYDloCpTHl21UoBewDgiIAXtA+IcAOwABgATFazL0LlC4nwcBStPpHb97aRjt9Vo7nTIaJteiILpQwFgoKpHUw+JQWVAHPpPSIgA)

### 💻 Code

```ts
interface Group {
    parentId: string | undefined
    //...
}

declare function fetchGroup(id: string): Promise<Group>

async function fetchAncestorGroups(leafId: string): Promise<Group[]> {
    const groups: Group[] = [];
    let nextId: string | undefined = leafId;

    while (nextId) {
        const group = await fetchGroup(nextId);
        //    ^^^^^ 'group' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. (7022)
        groups.push(group);
        nextId = group.parentId;
    }
    return groups.reverse();
}
```

### 🙁 Actual behavior

```
'group' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. (7022)
```

### 🙂 Expected behavior

No errors.

**Comment by RyanCavanaugh** [comment]
This really is a circularity; the fact that we can higher-order-reason our way out of it doesn't make it not a circularity. The core checking algorithm doesn't have mechanisms in place that would allow this to be resolved.

**Comment by disjukr** [comment]
https://github.com/microsoft/TypeScript/issues/30308 also related

**Comment by JoostK** [comment]
This is no longer reported as an error since TS 5.0, unlike [this scenario](https://github.com/microsoft/TypeScript/issues/33191#issuecomment-768406687).
